src/operations/blas2/gerp.hpp

```diff
@@ -0,0 +1,196 @@
+/***************************************************************************
+ *
+ *  @license
+ *  Copyright (C) Codeplay Software Limited
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ *  @filename gerp.hpp
+ *
+ **************************************************************************/
+
+#ifndef GERP_HPP
+#define GERP_HPP
+
+#include <operations/blas2_trees.h>
+#include <operations/blas_operators.hpp>
+#include <stdexcept>
+#include <vector>
+#include <views/view_sycl.hpp>
+
+namespace blas {
+
+// Row-Col index calculation for Row Major Lower Packed Matrix
+template <bool Single, bool isColMajor, bool isUpper, typename lhs_t,
+          typename rhs_1_t, typename rhs_2_t>
+template <int N>
+struct Gerp<Single, isColMajor, isUpper, lhs_t, rhs_1_t,
+            rhs_2_t>::compute_row_col<N, false, false> {
+  using index_t =
+      Gerp<Single, isColMajor, isUpper, lhs_t, rhs_1_t, rhs_2_t>::index_t;
+  using value_t =
+      Gerp<Single, isColMajor, isUpper, lhs_t, rhs_1_t, rhs_2_t>::value_t;
+  void operator()(const index_t id, const index_t, index_t& row, index_t& col) {
+    row = (-1 + cl::sycl::sqrt((value_t)(1 + 8 * id))) / 2;
```

Comment:
I don't like the use of a floating-point operation here. This comment applies everywhere you've used SYCL sqrt.

Maybe I've misunderstood quite how this works, but, if `value_t` is `float`, that means that we can have a roughly 1 million element matrix (1000 * 1000) before float is no longer able to represent (1 + 8 * 10000000) exactly. Then sqrt can have 2 ULP error, plus division of floating point can add more error. 

It looks like you test `n=1010` later on, so it might be that this is fine.

Reply:
Can you comment on whether there is a risk that `id` will be big enough for there to be issues with using floating point calculations - what about with half precision?

Reply:
Hi @hjabird, this issue is now fixed. I have tried running the benchmarks for upto `16384x16384` matrix size and the square roots are looking good (almost).